486 lucky money phase 2 update logic split money, add log and update validation

thanhphanvanncc opened the pull request:
PR: [#486](https://github.com/mezonai/mmn-tx-explorer/issues/723)

thanhphanvanncc committed:
#486 lucky money phase 2 update logic split money, add log and update validattion

hoangdohuy commented on dong-service/routes/routes.go:
```diff
@@ -94,7 +94,7 @@ func SetupRoutes(router *gin.Engine, cfg *config.Config) {
 		redEnvelopePrivate.GET("/detail/:id", redEnvelopeHandler.GetDetailRedEnvelopeByID)
 		redEnvelopePrivate.POST("/close-session", redEnvelopeHandler.CloseSessionRedEnvelope)
 		redEnvelopePrivate.GET("/claim-amount", redEnvelopeHandler.ClaimAmountRedEnvelope)
-		redEnvelopePrivate.POST("/:id/claim", redEnvelopeHandler.ClaimRedEnvelope)
```
why change endpoint?

phuongtrinhviet committed:
Merge develop into 486-lucky-money-phase-2-improvement-be - Keep refactored ExecuteClaim with 2-transaction approach

phuongtrinhviet committed:
resolve conflict

phuongtrinhviet committed:
486 lucky money phase 2 improvement be 2 (#645)

* #486 update logic to insert db first then call blockchain

* #486 delete old migration

* #486 update

* #486 updateClaimStatusInTx

phuongtrinhviet committed:
Merge branch 'develop' into 486-lucky-money-phase-2-improvement-be

phuongtrinhviet committed:
relsove conflict

sontung-ncc committed:
Merge branch 'develop' into 486-lucky-money-phase-2-improvement-be

phuongtrinhviet committed:
remove unnecessary request body

phongvu2812 committed:
remove unused var

phongvu2812 committed:
Merge pull request #725 from mezonai/486-lucky-money-phase-2-improvement-be-3

remove unnecessary request body

phongvu2812 committed:
Create 2 legacy api for sdk mmn-client-js

phongvu2812 committed:
fix flow create to support old api

phongvu2812 committed:
update legacy api

phongvu2812 committed:
Merge branch 'develop' into 486-lucky-money-phase-2-improvement-be

phongvu2812 committed:
Merge branch '486-lucky-money-phase-2-improvement-be' into lixi-be-v2

phongvu2812 committed:
update initializequeuelegacy for api legacy

phongvu2812 committed:
Merge branch 'develop' into 486-lucky-money-phase-2-improvement-be

phongvu2812 committed:
delete unused code

hoangdohuy committed:
Merge pull request #786 from mezonai/lixi-be-v2

Support both old and new flow of claim red-envelope

hoangdohuy commented on dong-service/migrations/021_delete_red_envelope_split_money_table.sql:
```diff
@@ -0,0 +1,4 @@
+-- Add status column to red_envelope_claim table for tracking claim status
```
đổi tên file migration đi với check số version xem khớp vs latest version vs develop chưa giúp a nhé

hoangdohuy commented on dong-service/repository/red_envelope_queue.go:
```diff
@@ -54,53 +55,161 @@ func (s *RedEnvelopeQueueService) InitializeRedEnvelope(redEnvelopeID string, to
+
+return 'OK'
+`)
+func (s *RedEnvelopeQueueService) AttemptClaimLegacy(redEnvelopeID string, userID int64) (int, error) {
```
TODO cho mấy hàm legacy giúp anh
`// TODO: should delete this legacy function in next release`

phongvu2812 committed:
Unified Redis Queue Service for both api routes

phongvu2812 commented on dong-service/handlers/red_envelope.go:
```diff
@@ -517,6 +488,57 @@ func (r *RedEnvelopeHandler) ClaimAmountRedEnvelopeQR(c *gin.Context) {
 
+	id := c.Query("id")
+	if validateClaimRedEnvelope := r.validateClaimRedEnvelope(c, id, userID); !validateClaimRedEnvelope {
+		return
```
trong hàm validateClaimRedEnvelope đã có phần logging ạ

phongvu2812 committed:
Merge branch 'develop' into 486-lucky-money-phase-2-improvement-be

hoangdohuy commented on dong-service/repository/red_envelope_repository.go:
```diff
@@ -905,60 +786,58 @@ func (r *RedEnvelopeRepository) ExecuteClaim(id, claimerWallet string, claimerUs
-		&claim.TransactionHash,
-	)
-
+	err = r.updateClaimStatusInTx(tx, claimID, constants.RedEnvelopeClaimStatusSuccess, &txHash)
```
Cần check xem nếu suggest nhưng vì 1 lí do nào đó, đoạn này fail thì có 1 cái cronjob nào quét các cbản ghi laim bị treo ở status là pending không?

phongvu2812 committed:
feat:
Using redis for red_evelope description
Update claim flow
Change type of split_id for legacy claim flow
Type casting for VerifyReservation output

phongvu2812 committed:
Red Envelope Split Money Migration to Redis

phongvu2812 committed:
Revert "Red Envelope Split Money Migration to Redis"

This reverts commit 0cf951a942a4d1f07b9cbbe097fb27e9ac10f715.

vietdangquoc commented on dong-service/repository/red_envelope_repository.go:
```diff
@@ -728,156 +736,189 @@ func (r *RedEnvelopeRepository) CloseSession(redEnvelopeID string, userID int64)
+			UPDATE %s.red_envelope
+			SET claimed_count = claimed_count - 1,
+				updated_at = $1
+			WHERE id = $2
```
AND claimed_count > 0
vì nế u laimed_count = claimed_count - 1, sẽ có error laimed_count < 0

phongvu2812 committed:
modified code base on comment

hoangdohuy commented on indexer/internal/storage/postgres.go:
```diff
@@ -3068,3 +3075,40 @@ func (p *PostgresConnector) failOfferStatus(
+		}
+
+		query := `
+			UPDATE dong_schema.red_envelope_claim
```
update trong loop thế này không được em ạ

hoangdohuy replied:
xem các hàm trước ấy sẽ query 1 lần lấy ra hết rồi validate bằng code, những transaction nào pass sẽ update ngược lại sau

hoangdohuy replied:
sender -> có phải là ví trung gian không.
receiver -> có phải là thằng nhận không.
amount có khớp với amount nó được chia không.

hoangdohuy commented on dong-service/repository/red_envelope_repository.go:
```diff
@@ -748,346 +734,215 @@ func (r *RedEnvelopeRepository) CloseSession(redEnvelopeID string, userID int64)
-	amount := types.NewBigIntString(claimAmount).Multiply(constants.TokenMultiplierBigIntString)
-	txHash, err = r.blockchainService.TransferMoney(walletInfo.EncryptedPrivateKey, envelope.RedEnvelopeWallet, claimerWallet, amount.String(), constants.TextDataLuckyMoney, constants.ExtraInfoLuckyMoney)
+	claimAmount := types.NewBigIntString(amount).Multiply(constants.TokenMultiplierBigIntString)
+	extraInfo := fmt.Sprintf(`{"type":"lucky-money","claim_id":%d}`, claimID)
```
+ constant không dùng raw string `lucky-money`

phongvu2812 committed:
update postfres.go
batch update and constant

QTR33 commented on dong-service/handlers/red_envelope.go:
```diff
@@ -517,52 +487,140 @@ func (r *RedEnvelopeHandler) ClaimAmountRedEnvelopeQR(c *gin.Context) {
 	}
 
+	var description string
+	description, err = r.repo.GetRedEnvelopeDescriptionByID(id)
```
done

phongvu2812 committed:
update validation in postgres.go

hoangdohuy reviewed (approved):
LGTM

phongvu2812 committed:
Merge branch 'develop' into 486-lucky-money-phase-2-improvement-be

phongvu2812 committed:
fix: change version migration to 33
